src/nextjs/utils/authMiddleware.ts

```diff
@@ -0,0 +1,89 @@
+import { NextMiddleware } from 'next/server';
+import { User, ApiError, createClient } from '@supabase/supabase-js';
+import { CookieOptions } from '../types';
+import { COOKIE_OPTIONS } from '../../shared/utils/constants';
+import { jwtDecoder } from '../../shared/utils/jwt';
+
+export const authMiddleware =
+  (
+    /**
+     * Path relative to the site root to redirect an
+     * unauthenticated visitor.
+     *
+     * The original request route will be appended via
+     * a `ret` query parameter, ex: `?ret=%2Fdashboard`
+     */
+    redirectTo = '/',
+    cookieOptions: CookieOptions = COOKIE_OPTIONS
+  ): NextMiddleware =>
+  async (req) => {
+    try {
+      if (
+        !process.env.NEXT_PUBLIC_SUPABASE_URL ||
+        !process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY
+      ) {
+        throw new Error(
+          'NEXT_PUBLIC_SUPABASE_URL and NEXT_PUBLIC_SUPABASE_ANON_KEY env variables are required!'
+        );
+      }
+      if (!req.cookies) {
+        throw new Error('Not able to parse cookies!');
+      }
+      const supabase = createClient(
+        process.env.NEXT_PUBLIC_SUPABASE_URL,
+        process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY
+      );
+      const access_token = req.cookies[`${cookieOptions.name!}-access-token`];
+      const refresh_token = req.cookies[`${cookieOptions.name!}-refresh-token`];
+
+      const getUser = async (): Promise<{
+        user: User | null;
+        error: ApiError | null;
+      }> => {
+        if (!access_token) {
+          throw new Error('No cookie found!');
+        }
+        // Get payload from access token.
+        const jwtUser = jwtDecoder(access_token);
+        if (!jwtUser?.exp) {
+          throw new Error('Not able to parse JWT payload!');
+        }
+        const timeNow = Math.round(Date.now() / 1000);
+        if (jwtUser.exp < timeNow) {
+          if (!refresh_token) {
+            throw new Error('No refresh_token cookie found!');
+          }
+          const res = await supabase.auth.api.refreshAccessToken(refresh_token);
+          return { user: res.data?.user ?? null, error: res.error };
+        }
+        const res = await supabase.auth.api.getUser(access_token);
+        return { user: res.user, error: res.error };
+      };
+
+      const authResult = await getUser();
+
+      if (authResult.error) {
+        throw new Error(
+          `Authorization error, redirecting to login page: ${authResult.error.message}`
+        );
+      } else if (!authResult.user) {
+        throw new Error('No auth user, redirecting');
+      }
+
+      // Authentication successful, forward request to protected route
+      return NextResponse.next();
+    } catch (err: unknown) {
+      if (err instanceof Error) {
+        console.log(
+          `Could not authenticate request, redirecting to ${redirectTo}:`,
+          err
+        );
+      }
+      const url = req.nextUrl.clone();
+      url.pathname = redirectTo;
+      url.searchParams.set(`ret`, req.nextUrl.pathname);
```

Comment:
You okay to call this param `nextUrl` instead?

Reply:
No problem, name here is arbitrary. Perhaps `redirectUrl` would be more descriptive?

Because we're constructing a new URL by taking the base of the existing nextUrl (ie: `https://localhost:3000` or `https://example.com`) and replacing it's path (`/dashboard` with `/login`), appending as a query parameter the path the user is bounced from.

Reply:
Yah, I like `redirectUrl` 👍

Reply:
Or could be `redirectTo` also I guess.